### ISSUE TITLE
Allow outbound passthrough on reserved ports

### ISSUE THREAD

**Issue by ricec** [istio/istio#36040]
**Describe the feature request**

Istio currently blocks outbound traffic to external systems on a subset of its reserved ports when traffic is proxied through the sidecar using ALLOW_ANY outbound traffic policy. For example, if a workload would like to access an external system on port 15021, this traffic is currently blocked by the sidecar.

In the interest of making the proxy as transparent as possible, we should do away with these restrictions/conflicts by reducing the scope of listeners down to the minimal set of IP addresses needed (e.g. instead of 0.0.0.0, listen on 127.0.0.1 and pod_ip). Please see the "Additional Context" for specific proposals per reserved port. 

**Describe alternatives you've considered**

- Create ServiceEntry configurations for each external service that uses an Istio reserved port
  - This creates Envoy listeners that will be prioritized above the default reserved-port listeners due to address specificity
  - This is not optimal, since many operators will use ALLOW_ANY specifically to avoid having to maintain service entries for external systems. Additionally, this is cumbersome when destination systems leverage port ranges as opposed to single, well-defined ports.
- Bypass the proxy for systems using reserved ports
  - This is not optimal, since it means traffic to those systems will be excluded from the benefits of the service mesh
- Reconfigure destination systems to use other ports
  - This is sometimes possible, sometimes not, depending on system ownership and configurability of the destination

**Affected product area (please put an X in all that apply)**

[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[ ] Multi Cluster
[ ] Virtual Machine
[ ] Multi Control Plane

**Additional context**

I have done some analysis on the changes required for this:

- 15000
  - Currently listens only on localhost https://github.com/istio/istio/blob/452101a4d549e580bcf17e4b6ef053e968a52c2b/tools/packaging/common/envoy_bootstrap.json#L153
  - No change needed
- 15001
  - Currently blackholes any traffic with original destination on port 15001 https://github.com/istio/istio/blob/da6178604559bdf2c707a57f452d16bee0de90c8/pilot/pkg/networking/core/v1alpha3/listener_builder.go#L744-L753
  - **Proposal:** Leverage sourceType in the FilterChainMatch to scope the blackhole filter chain down to only traffic coming from the proxy back to itself on 15001
- 15004
  - Currently listens only on localhost https://github.com/istio/istio/blob/452101a4d549e580bcf17e4b6ef053e968a52c2b/pkg/istio-agent/xds_proxy.go#L839-L840
  - No change needed
- 15006
  - Currently blackholes any traffic with original destination on port 15006 https://github.com/istio/istio/blob/da6178604559bdf2c707a57f452d16bee0de90c8/pilot/pkg/networking/core/v1alpha3/listener_builder.go#L558-L570 
  - **Proposal:** Leverage sourceType in the FilterChainMatch to scope the blackhole filter chain down to only traffic coming from the proxy back to itself on 15006
- 15008
  - Couldn't find an actual listener for this. Appears to have been preemptively added as part of https://github.com/istio/istio/issues/24540 via https://github.com/istio/istio/pull/25392 
  - Seems no change needed for now, but would need to ensure reserved ports introduced as part of "Better transport security for HTTP" are unobtrusive
- 15009
  - Like 15008, appears to be WIP, part of https://github.com/istio/istio/issues/24540 via https://github.com/istio/istio/pull/28244
  - Seems no change needed for now, but would need to ensure reserved ports introduced as part of "Better transport security for HTTP" are unobtrusive
- 15020
  - Listens on 0.0.0.0. However, this is not an Envoy listener, therefore, there shouldn't be any bad interactions when there is passthrough traffic targeting this port on an external system. https://github.com/istio/istio/blob/c017654753cd8f4254abbe0567ff9c728d2034d2/pilot/cmd/pilot-agent/status/server.go#L317
  - Note: It appears that with istio-init, this port is redirected to the proxy on OUTPUT (seems correct), but on CNI, it is excluded from redirect/proxy on OUTPUT (seems incorrect) https://github.com/istio/istio/blob/437970b95b6a67b1df4d2211095f1e0ef3f9a9e6/cni/pkg/plugin/redirect.go#L32-L38 https://github.com/istio/istio/blob/437970b95b6a67b1df4d2211095f1e0ef3f9a9e6/cni/pkg/plugin/redirect.go#L63
  - **Proposal:** Remove the outbound exclusion for this port in CNI iptables rules so outbound traffic on 15020 will be redirected to 15001 for passthrough
- 15021
  - Currently listens on 0.0.0.0. Traffic w/ original dst on port 15021 will be redirected to 15001. That listener will delegate to this one, preventing outbound passthrough to external services on 15021. https://github.com/istio/istio/blob/452101a4d549e580bcf17e4b6ef053e968a52c2b/tools/packaging/common/envoy_bootstrap.json#L520
  - **Proposal:** Instead of one listener on 0.0.0.0, let's split this into 2: 1 listening on pod_ip and 1 listening on localhost. This will not match traffic to external services, thus allowing it to passthrough.
- 15053
  - Currently listens on localhost https://github.com/istio/istio/blob/dc98beef00c9b851ee8c79e6dfb89fea8f19c64a/pilot/cmd/pilot-agent/options/options.go#L94
  - No change needed
- 15090
  - Currently listens on 0.0.0.0. Traffic w/ original dst on port 15090 will be redirected to 15001. That listener will delegate to this one, preventing outbound passthrough to external services on 15090. https://github.com/istio/istio/blob/452101a4d549e580bcf17e4b6ef053e968a52c2b/tools/packaging/common/envoy_bootstrap.json#L471
  - **Proposal:** Instead of one listener on 0.0.0.0, let's split this into 2: 1 listening on pod_ip and 1 listening on localhost. This will not match traffic to external services, thus allowing it to passthrough.

**Proof of Concept**

Additionally, I have put together a proof-of-concept implementation of the above proposals to be further developed if this approach makes sense: https://github.com/ricec/istio/commit/14f7ff4b25723215c28e3879a0b17fd61afa2ab6

**Comment by ricec** [comment]
@howardjohn , @lambdai, if I can get your buy-in on the proposed approach, I'd be happy to further develop the POC code (ricec@14f7ff4) and submit a PR. Will you please take a look?

**Comment by ricec** [comment]
@howardjohn , @lambdai, any thoughts?

**Comment by howardjohn** [comment]
Thanks for the POC - appreciate the effort here. I agree with the motivation - allow traffic to all ports. I do have some concerns.

Mixing inbound with outbound is a *huge* source of vulnerability. I don't know if I would ever feel comfortable with having the ability to send outbound traffic on the inbound path. The issue that can come up - and has, in the past, is I send a request to `foo` and convince it to take the outbound path instead of the inbound, effectively a SSRF.

The other concern is if/when we fix this, we still do not reliable have "send traffic to all ports" (a la https://github.com/istio/istio/issues/23983). So what may happen is even if this is fixed we still run into issues. In the best case, we are hitting PassthroughCluster so we get minimal visibility. In the worst case, there is a TCP ServiceEntry without `address` set which captures all traffic on that port. Another case is that some ports will have protocol sniffing enabled, which can have different behavior on different ports.

Coincidentally I have been working on thinking about how we can refactor our outbound routing logic that would address all of these issues. I think this combined with some of the ideas you have above should be able to satisfy these cases

**Comment by ricec** [comment]
Thanks for the feedback @howardjohn. The SSRF concern definitely makes sense. While I don't think that would be possible in this case (I could be wrong), the POC code for the inbound listener certainly would make it easier for a such a vulnerability to sneak it's way in as the code evolves over time. Either way, in retrospect, it definitely violates the principle of least astonishment and is my least favorite part of this proposal :)

What about this as an alternative... We could restrict the virtual inbound listener to only listen on the pod's IP instead of 0.0.0.0. That way, this listener will not match outbound traffic to external systems. I think I had originally shied away from this idea because I thought we would need to listen on both pod IP and localhost, and I didn't want to propose duplicating the virtual inbound listener. As I'm considering it now, though, I can't think of scenarios where the virtual inbound listener would need to listen on localhost. What do you think?

Regarding the other concerns, I agree there are still potential issues that could crop up when relying on the implicit ALLOW_ANY behavior as opposed to explicitly defining external services, but most of what you've listed seems avoidable. And even if we were able to specify port ranges / wildcards in service entries, we would still run into issues with these conflicting listeners. For example, if I create a service entry on port 15006, but I use `resolution: DNS` as opposed to listing out `addresses`, the resulting listener is not specific enough to be matched ahead of the virtual inbound listener, and outbound connectivity to that service remains broken. Point is, I think that addressing these port conflicts is a step forward for both ALLOW_ANY and REGISTRY_ONLY.

I'm definitely interested in understanding what you have in mind for redesigning the outbound routing logic, so if there's an open issue tracking it, please do share. And let me know if/how I can help integrate some of this proposal into it.

**Comment by howardjohn** [comment]
> As I'm considering it now, though, I can't think of scenarios where the virtual inbound listener would need to listen on localhost. What do you think?

It might _only_ need to listen on localhost? I think iptables REDIRECT sends to localhost. I could be wrong though. That seems like a plausible fix here.

Will try to post the proposal later today

**Comment by ricec** [comment]
So this ended up being slightly more complicated as I built out the necessary changes. Here's what I found:

- For all traffic inbound to the pod from external, the REDIRECT from PREROUTING will result in Pod IP being the destination
- For all traffic outbound from the pod, the REDIRECT from OUTPUT will result in localhost being the destination
- If we _only_ listen on localhost, inbound traffic from external will break
- If we _only_ listen on pod IP, outbound traffic from app to Service VIP will break (app -> envoy (outbound) -> envoy (inbound) -> app)
- The same general behavior applies in TPROXY mode as well

So this throws a wrench in the simple approach of only listening on localhost or pod IP for inbound. To address this, I made some modifications to the iptables rules:
- For REDIRECT mode, I changed the target for ISTIO_IN_REDIRECT to be `-j DNAT --to-destination pod_ip:15006`
  - SO_ORIGINAL_DST works with DNAT as well, so this approach doesn't break anything
- For TPROXY mode, I set `--to-ip pod_ip` on the rule with the TPROXY target

Net effect is that traffic destined to the inbound listener will always have pod IP as destination, so we can then proceed with limiting the virtual inbound listener to only listen on Pod IP. The virtual outbound listener can safely listen only on localhost. This _significantly_ reduces the potential cases in which outbound traffic would end up on the inbound listener and vice versa, which I think is a good thing.

Almost all tests are passing locally at this point (will fix up the rest next week). @howardjohn, could you take a look at #37371, let me know what you think, and give it an "ok-to-test", please? Sorry it's such a huge PR, but most of the changes are just goldens being refreshed.

**Comment by nagavijayan-nagarathinam** [comment]
I have used this annotation to fix this issue. but does it make any side effect in service mesh network ? `traffic.sidecar.istio.io/excludeOutboundPorts: 15001`
